File: Tests/AINotebookCoreTests/ChunkerTests.swift
import XCTest
@testable import AINotebookCore

final class ChunkerTests: XCTestCase {
    func testShortTextProducesSingleChunk() {
        let drafts = Chunker.chunk("Hello world.")
        XCTAssertEqual(drafts.count, 1)
        XCTAssertEqual(drafts[0].text, "Hello world.")
        XCTAssertGreaterThan(drafts[0].tokenCount, 0)
    }

    func testEmptyOrWhitespaceProducesNoChunks() {
        XCTAssertTrue(Chunker.chunk("").isEmpty)
        XCTAssertTrue(Chunker.chunk("   \n\t ").isEmpty)
    }

    func testLongTextSplitsIntoMultipleChunks() {
        let para = String(repeating: "word ", count: 2000)  // ~10 000 chars
        let drafts = Chunker.chunk(para)
        XCTAssertGreaterThan(drafts.count, 1)
        // Every chunk under the hard cap (2 048 chars + a small slack for
        // not breaking mid-word).
        for d in drafts {
            XCTAssertLessThanOrEqual(d.text.count, 2_100, "chunk too big: \(d.text.count)")
        }
    }

    func testChunksOverlap() {
        let para = String(repeating: "word ", count: 2000)
        let drafts = Chunker.chunk(para)
        guard drafts.count >= 2 else { return XCTFail("need at least 2 chunks") }
        // Last 200 chars of chunk N appear at the start of chunk N+1
        // (because of the 256-char overlap window — allow some boundary slack).
        let tail = String(drafts[0].text.suffix(200))
        XCTAssertTrue(
            drafts[1].text.contains(tail.prefix(100)),
            "expected overlap between consecutive chunks"
        )
    }

    func testWindowAndOverlapAreOverridable() {
        let drafts = Chunker.chunk(
            String(repeating: "a ", count: 500),
            windowChars: 200,
            overlapChars: 50
        )
        XCTAssertGreaterThan(drafts.count, 3)
        for d in drafts {
            XCTAssertLessThanOrEqual(d.text.count, 220)
        }
    }
}

File: windows/tests/AINotebook.Core.Tests/Ingestion/ChunkerTests.cs
using AINotebook.Core.Ingestion;
using Xunit;

namespace AINotebook.Core.Tests.Ingestion;

public class ChunkerTests
{
    [Fact]
    public void ShortTextProducesSingleChunk()
    {
        var drafts = Chunker.Chunk("Hello world.");
        Assert.Single(drafts);
        Assert.Equal("Hello world.", drafts[0].Text);
        Assert.True(drafts[0].TokenCount > 0);
    }

    [Fact]
    public void EmptyOrWhitespaceProducesNoChunks()
    {
        Assert.Empty(Chunker.Chunk(""));
        Assert.Empty(Chunker.Chunk("   \n\t "));
    }

    [Fact]
    public void LongTextSplitsIntoMultipleChunks()
    {
        var para = string.Concat(Enumerable.Repeat("word ", 2000)); // ~10 000 chars
        var drafts = Chunker.Chunk(para);
        Assert.True(drafts.Count > 1);
        // Every chunk under the hard cap (2 048 chars + small slack for not breaking mid-word).
        foreach (var d in drafts)
        {
            Assert.True(TextElements.Count(d.Text) <= 2_100, $"chunk too big: {TextElements.Count(d.Text)}");
        }
    }

    [Fact]
    public void ChunksOverlap()
    {
        var para = string.Concat(Enumerable.Repeat("word ", 2000));
        var drafts = Chunker.Chunk(para);
        Assert.True(drafts.Count >= 2, "need at least 2 chunks");
        // Last 200 chars of chunk N appear at the start of chunk N+1 (256-char overlap window).
        var tail = new string(drafts[0].Text.AsSpan(drafts[0].Text.Length - 200).ToArray());
        Assert.Contains(tail[..100], drafts[1].Text);
    }

    [Fact]
    public void WindowAndOverlapAreOverridable()
    {
        var drafts = Chunker.Chunk(
            string.Concat(Enumerable.Repeat("a ", 500)),
            windowChars: 200,
            overlapChars: 50);
        Assert.True(drafts.Count > 3);
        foreach (var d in drafts)
        {
            Assert.True(TextElements.Count(d.Text) <= 220);
        }
    }
}
